### ISSUE TITLE
HystrixObservableCommand - Fallback via a network call, Thread-pool isolation?

### ISSUE THREAD

**Issue by gt-tech** [Netflix/Hystrix#1510]
There are a few "closed" issues that indicates the good reasoning behind why thread-pool isolation was disabled/removed from HystrixObservableCommand. However I didn't see any clear instructions/reference on how would a Fallback (via another command, with network call) work.

Observable Command A -> Makes a network call to remote service
Observable Command B -> Makes a network call to stale cache (fallback for Command A failure)

In case of HystrixCommand (non-observable), framework ensured that fallback executed as an independent HystrixCommand since it was invoked on caller's thread (not on Command A's threadpool) and hence independent threadpool settings etc. were declaratively possible for Command A and Command B.

However, in case of both commands being Observable, since Observable command are using "semaphores", essentially utilizing caller thread, how can we configure it such that fallback does get execute in a separate thread?

One way (or perhaps only way?) would be to not configure a fallback in Command A and let the exception reach to user/caller of command A which can then use one of OnError* operators from RxJava to execute Command B, is there any other way that would be more cleaner?

**Comment by mattrjacobs** [comment]
To be clear, it's still possible, though not recommended.  I haven't had the time to make that pattern impossible yet.

Even in thread-isolation, fallbacks can happen in a variety of places:
* On normal error - in the Hystrix thread
* On open circuit/rejection - in the caller thread
* On timeout - in the Timer thread

In semaphore isolation, all execution, including fallbacks, happens on the caller thread.  In the case of `HystrixObservableCommand`, you're providing `Observable`s.  So you have the ability to move those around to threads of your choosing.

For example, you can implement `runWithFallback` as `Observable.defer(go-do-some-work).subscribeOn(Schedulers.io())`.  The `go-do-some-work` method (and all subsequently-chained work) will happen on an RxIoScheduler thread.  

In general, you can use `subscribeOn`/`observeOn` to have very fine-grained control over where work occurs.

Hope that helps!

**Comment by gt-tech** [comment]
Thanks Matt, I am aware of what can be done on calling code (which invokes Command A) to use appropriate threads for fallback Command B, but all of those possibilities only exist if Command A didn't have a fallback because if it did, then technically hystrix will ensure to call fallback (Command B) in same thread that executed Command A and never give a chance for caller to use different threads.

In other words, it seems like one shouldn't use fallback commands in case async use of hystrix if they want to use different thread-pools for main execution and fallback?

**Comment by gt-tech** [comment]
On second thoughts, I think i got it what you meant, we can still keep control of thread on which fallback execute, we would just need to do either of following:
1) don't give fallback to primary command and let error reach back to caller which can then execute fallback using thread of choice
2) give a fallback to primary command that's already wrapped up in Observable with appropriate scheduler.

I will give it a shot but it makes sense now. thanks
